### PR TITLE
feat(bazel): Introduce standalone Bazel build graph extractor via Aspect

### PR DESCRIPTION
**Pull request by neilu-jain**
## Description
This PR introduces a brand new, highly robust standalone extractor for Bazel environments (`bazel/aspect`). 

While `osv-scalibr` already supports Bzlmod lockfile parsing, there are many legacy workspaces, dynamically generated targets (like `gazelle` go_deps), and complex internal `http_archive` rules that are completely opaque to static lockfile parsers.

This new plugin bridges that gap by running a custom Bazel aspect (`.scalibr_aspect.bzl`) directly against the live Bazel graph (`bazel build --aspects=... //...`). This allows Scalibr to deeply traverse and extract canonical dependency metadata from any Bazel workspace, regardless of how the dependencies are structured.

### Key Features
* **Bzlmod Native Compatibility:** Accurately extracts canonical repository names from complex Bzlmod extension mangling (e.g. `gazelle~go_deps~com_github_foo` -> `com_github_foo`).
* **Fallback Version Extraction:** Implements generic SemVer heuristics to pull versions out of legacy `WORKSPACE` `http_archive` URLs and `strip_prefix` attributes where explicit versions do not exist.
* **Highly Performant:** Regex compilations are hoisted, allocations are optimized, and output package lists are deterministically sorted to prevent flaky CI diffs.
* **Comprehensive Test Suite:** Includes a robust `_test.go` suite asserting edge cases for URL parsing, Bzlmod naming, and aspect extraction.
* **Fully Audited:** Code strictly adheres to `golangci-lint`, `revive`, and `osv-scalibr` architectural standards.

### Testing
Tested against large-scale monolithic Bazel repositories. It successfully extracted thousands of dependencies (including `rules_go`, `rules_rust`, and `pkg:generic` tarballs) which successfully mapped to hundreds of real CVEs via `osv-scanner`, capturing 100% of the vulnerability footprint compared to bespoke internal scanners.

### CODE REVIEW

File: extractor/standalone/bazel/aspect/aspect.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+// Name returns the extractor's name.
+func (e *Extractor) Name() string { return Name }
+
+// Version returns the extractor's version.
+func (e *Extractor) Version() int { return 0 }
+
+// Requirements returns the requirements for this extractor.
+func (e *Extractor) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{RunningSystem: true, DirectFS: true}
```
Comment (erikvarga) [inline extractor/standalone/bazel/aspect/aspect.go]: Since we're executing commands directly, can a malicious attacker-controlled artifact cause RCE on the system running this scanner? If so we should add AllowUnsafePlugins: true

File: extractor/standalone/list/list.go [hunk 2]
```diff
@@ -74,6 +80,7 @@ var (
 		"extractors/all":     vals(All),
 		"all":                vals(All),
 		"containers":         vals(Containers),
+		"bazel":              vals(Bazel),
```
Comment (erikvarga) [inline extractor/standalone/list/list.go]: Also add this to docs/supported_inventory_types.md

File: extractor/standalone/bazel/aspect/aspect.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+		return inventory.Inventory{}, errors.New("ScanRoot is required")
+	}
+
+	_, err := exec.LookPath("bazel")
+	if err != nil {
+		return inventory.Inventory{}, errors.New("bazel not found in PATH")
+	}
+
+	// Setup a temporary workspace package for the aspect
+	aspectDir := filepath.Join(input.ScanRoot.Path, ".scalibr_aspect")
```
Comment (erikvarga) [inline extractor/standalone/bazel/aspect/aspect.go]: How about os.CreateTemp?

File: extractor/filesystem/bazel/aspect/aspect.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+	// target is the Bazel target to run the aspect on.
+	target string
+	// keepGoing determines whether to use the --keep_going flag.
+	keepGoing bool
+}
+
+// New returns a new instance of the Extractor.
+func New(_ *cpb.PluginConfig) (standalone.Extractor, error) {
+	e := &Extractor{
+		target:    "//...",
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect.go:76]: We should configure these from the PluginConfig and use these as default values.

File: extractor/filesystem/bazel/aspect/aspect.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+	"github.com/google/osv-scalibr/plugin"
+)
+
+//go:embed scalibr_aspect.bzl
+var scalibrAspectBzl []byte
+
+// Name is the unique name of this extractor.
+const Name = "bazel/aspect"
+
+// Extractor is a standalone extractor for Bazel dependencies using an aspect.
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect.go]: I'm a bit confused about why we picked a standalone extractor

Filesystem extractors iterate over the filesystem and extract files that are relevant for them. Standalone extractors are usually used for non-filesystem related data like windows registry values.

In this case it looks like you're running blaze on a target - does that always reside at the scan root?

**Review by neilu-jain — COMMENTED** [review]
@erikvarga updated!

File: extractor/standalone/bazel/aspect/aspect.go (hunk 1 above)
Comment (neilu-jain) [reply extractor/standalone/bazel/aspect/aspect.go]: Done!

File: extractor/standalone/list/list.go (hunk 2 above)
Comment (neilu-jain) [reply extractor/standalone/list/list.go]: Done!

File: extractor/standalone/bazel/aspect/aspect.go (hunk 3 above)
Comment (neilu-jain) [reply extractor/standalone/bazel/aspect/aspect.go]: Done!

File: extractor/filesystem/bazel/aspect/aspect.go (hunk 4 above)
Comment (neilu-jain) [reply extractor/filesystem/bazel/aspect/aspect.go:76]: Done!

File: extractor/filesystem/bazel/aspect/aspect.go (hunk 5 above)
Comment (neilu-jain) [reply extractor/filesystem/bazel/aspect/aspect.go]: It seems that a `filesystem` extractor runs its extraction logic repeatedly every time it matches a target file format. Since we are querying the target `//...` (which natively evaluates the entire workspace tree in one pass), or a cinfigured target - running it as a filesystem extractor would trigger a massive redundant Bazel build for every single `WORKSPACE` or `BUILD` file it encounters. 

By using a `standalone` extractor, we execute the Bazel aspect exactly once against the `ScanRoot`, capturing the entire dependency graph perfectly in a single execution.

Comment (erikvarga) [reply extractor/filesystem/bazel/aspect/aspect.go]: So in this case a regular SCALIBR scan on a workstation (i.e. not source repo) would trigger bazel build on "/" on the given workstation. Do we foresee any issues with that? Apart from performance, I'm a bit concerned we'll skip all the blocklists normally configured for the filesystem extraction and traverse through every directory, even ones that the user might have wanted to exclude.

Comment (neilu-jain) [reply extractor/filesystem/bazel/aspect/aspect.go]: To prevent rogue execution, I've pushed a commit that adds an explicit `isBazelWorkspace()` validation check. Before the extractor invokes the `bazel` CLI, it ensures `WORKSPACE` or `MODULE.bazel` exists directly at the `ScanRoot`. If not, it safely assumes this is an arbitrary filesystem directory and gracefully returns an empty inventory. This guarantees we won't accidentally trigger massive unconstrained Bazel builds on standard workstation scans.

Regarding blocklists (e.g. `--skip-dirs`): Because Bazel requires a complete and valid dependency graph to execute its analysis phase, we cannot natively map filesystem exclusions into Bazel. If a user excludes a directory that the Bazel graph fundamentally depends on, the entire `bazel build` command would fatally crash and fail to extract any dependencies at all. 

Instead, if users want to exclude specific packages from the Bazel scan, they should use the new `BazelAspectConfig` message we just added to override the default `//...` target to something more scoped (e.g. `//src/...`). This natively isolates the Bazel evaluation without breaking the build graph.

File: extractor/standalone/bazel/aspect/aspect.go (hunk 1 above)
Comment (erikvarga) [reply extractor/standalone/bazel/aspect/aspect.go]: Let's also add an explanation for why this plugin is unsafe

File: extractor/filesystem/bazel/aspect/aspect.go (hunk 5 above)
Comment (erikvarga) [reply extractor/filesystem/bazel/aspect/aspect.go]: So it sounds like we'd only want the command to run from the start of Bazel workspaces. SCALIBR is supposed to find all inventories on a given host even when they're deeper down, so returning empty if the scan isn't started from a bazel workspace isn't that great either.

I think we could remediate that in a filesystem extractor while avoiding duplicate checks: Store a list of directories that have already been processed in the extractor fields. Then we can make IsRequired look something like

func Isrequired(path) {
  return isBazelWorkspace(path) && !directoryProcessed(path)
}

This way we'd run the bazel command just once for every distinct code repo found on the filesystem. For source code repos this would only be done once since root dir contains the WORKSPACE file and thus entire scan root is marked as processed after the first command invocation.

File: extractor/filesystem/bazel/aspect/aspect.go [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+	// keepGoing determines whether to use the --keep_going flag.
+	keepGoing bool
+	// processed tracks workspace roots that have already been processed to avoid duplicate executions.
+	processed sync.Map
+}
+
+// New returns a new instance of the Extractor.
+func New(cfg *cpb.PluginConfig) (filesystem.Extractor, error) {
+	e := &Extractor{
+		target:    "//...",
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect.go:76]: When Extract() is called on a given directory, would the command that receives this run under the OS root? Or the workspace root?

File: extractor/filesystem/bazel/aspect/scalibr_aspect.bzl
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/scalibr_aspect.bzl:1]: I couldn't quite find how this file is used - Is it loaded by the SCALIBR Go binary in some way? Does the bazel extractor expect SCALIBR source to be present in order to be able to run the scan?

**Comment by erikvarga** [comment]
AFAICS the businesslogic looks okay now - Note the linter errors though.

Let's also add some unit tests for the new extractor introduced

File: extractor/filesystem/bazel/aspect/aspect.go (hunk 6 above)
Comment (neilu-jain) [reply extractor/filesystem/bazel/aspect/aspect.go:76]: The bazel command is executed directly under the workspace root (the directory where the WORKSPACE or MODULE.bazel file was discovered by the filesystem traversal). We set the Dir field on the exec.Cmd to that specific path, so it never runs from the OS root.

File: extractor/filesystem/bazel/aspect/scalibr_aspect.bzl
Comment (neilu-jain) [reply extractor/filesystem/bazel/aspect/scalibr_aspect.bzl:1]: The .scalibr_aspect.bzl file is bundled directly into the compiled SCALIBR binary using //go:embed. At runtime, the extractor writes this embedded file out to a temporary directory so the Bazel CLI can reference it. Because of this, it does not expect the SCALIBR source tree to be present on the target host in order to run the scan.

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+			}
+			got := e.FileRequired(simplefileapi.New(tt.inputPath, nil))
+			if got != tt.want {
+				t.Errorf("FileRequired(%s) got = %v, want %v", tt.inputPath, got, tt.want)
+			}
+		})
+	}
+}
+
+func TestExtractor_NameVersion(t *testing.T) {
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go]: We can skip this and Requirements since it's just repeating the implementation

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+		t.Fatalf("aspect.New() error: %v", err)
+	}
+	reqs := e.Requirements()
+	if !reqs.RunningSystem {
+		t.Errorf("Requirements().RunningSystem = %v, want %v", reqs.RunningSystem, true)
+	}
+	if !reqs.DirectFS {
+		t.Errorf("Requirements().DirectFS = %v, want %v", reqs.DirectFS, true)
+	}
+}
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go:410]: Extract() is the function with the main businesslogic that should be tested. We should add tests for it.

Comment (erikvarga) [reply extractor/filesystem/bazel/aspect/aspect_test.go:410]: You'll likely have to mock out the command runs in order to test it. I'm not sure if Go provides built-in ways to do that but you can for example pass in a custom command runner class to the extractor that defaults to the "os/exec" implementation and override that with a mock exec class in the tests.

Comment (neilu-jain) [reply extractor/filesystem/bazel/aspect/aspect_test.go:410]: Done! Introduced a CommandRunner interface and a NewWithRunner constructor (defaulting to standard os/exec in production).

Added comprehensive table-driven tests for Extract() in aspect_test.go using a mock runner, verifying:

Successful aspect execution and parsing of the BEP event stream
PURL normalization and cross-ecosystem package mapping
Target and --keep_going overrides from PluginConfig
Error handling (missing bazel binary, invalid JSON, etc.)
Workspace skipping and deduplication across multiple traversal hits

**Comment by neilu-jain** [comment]
@erikvarga

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 9]
```diff
@@ -59,29 +96,396 @@ func TestExtractor_FileRequired(t *testing.T) {
...
+			}
+
+			if !tt.wantErr {
+				if diff := cmp.Diff(tt.wantInv, got, cmpopts.EquateEmpty()); diff != "" {
+					t.Errorf("Extract() inventory mismatch (-want +got):\n%s", diff)
+				}
+
+				// If testing already-processed scenario, call Extract again
+				if tt.name == "workspace already processed" {
+					got2, err2 := e.Extract(context.Background(), &filesystem.ScanInput{
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go]: t.Context()

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 10]
```diff
@@ -59,29 +96,396 @@ func TestExtractor_FileRequired(t *testing.T) {
...
-	if !reqs.RunningSystem {
-		t.Errorf("Requirements().RunningSystem = %v, want %v", reqs.RunningSystem, true)
+	tests := []struct {
+		name         string
+		setupFs      func(t *testing.T, wsDir string)
+		pluginConfig *cpb.PluginConfig
+		mockRunner   func(t *testing.T, wsDir string) aspect.CommandRunner
+		scanPath     string
+		wantInv      inventory.Inventory
+		wantErr      bool
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go]: use err here and cmp.Diff to check for it - https://github.com/google/osv-scalibr/blob/73d42b5363c42ad35e9ff91caa7b0a29e5d5b269/scalibr_test.go#L737

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 11]
```diff
@@ -59,29 +96,396 @@ func TestExtractor_FileRequired(t *testing.T) {
...
+	}{
+		{
+			name: "success with mixed dependency types and deduplication",
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go]: use underscores in test names instead of spaces

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 12]
```diff
@@ -59,29 +96,396 @@ func TestExtractor_FileRequired(t *testing.T) {
...
+							if after, ok := strings.CutPrefix(arg, "--build_event_json_file="); ok {
+								bepPath = after
+							}
+						}
+						if bepPath == "" {
+							t.Fatal("missing --build_event_json_file argument")
+						}
+
+						// Create mock aspect output JSON files
+						aspectOutputs := []aspectTestData{
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go:160]: Do we need this many mock outputs? Can one or two replicate the same test cases?

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 13]
```diff
@@ -59,29 +96,396 @@ func TestExtractor_FileRequired(t *testing.T) {
...
+				tt.setupFs(t, wsDir)
+			}
+
+			cfg := tt.pluginConfig
+			if cfg == nil {
+				cfg = &cpb.PluginConfig{}
+			}
+
+			var runner aspect.CommandRunner
+			if tt.mockRunner != nil {
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go]: from what I can see this is never nil

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 14]
```diff
@@ -59,29 +96,396 @@ func TestExtractor_FileRequired(t *testing.T) {
...
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			wsDir := t.TempDir()
+			if tt.setupFs != nil {
+				tt.setupFs(t, wsDir)
+			}
+
+			cfg := tt.pluginConfig
+			if cfg == nil {
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go]: What happens if cfg is nil? I'd have assumed all its getter methods are nil-resistant.

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 15]
```diff
@@ -59,29 +96,396 @@ func TestExtractor_FileRequired(t *testing.T) {
...
+			if tt.mockRunner != nil {
+				runner = tt.mockRunner(t, wsDir)
+			}
+
+			e, err := aspect.NewWithRunner(cfg, runner)
+			if err != nil {
+				t.Fatalf("aspect.NewWithRunner() error: %v", err)
+			}
+
+			got, err := e.Extract(context.Background(), &filesystem.ScanInput{
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go]: t.Context()

File: extractor/filesystem/bazel/aspect/aspect_test.go [hunk 16]
```diff
@@ -59,29 +96,396 @@ func TestExtractor_FileRequired(t *testing.T) {
...
+			if (err != nil) != tt.wantErr {
+				t.Fatalf("Extract() error = %v, wantErr %v", err, tt.wantErr)
+			}
+
+			if !tt.wantErr {
+				if diff := cmp.Diff(tt.wantInv, got, cmpopts.EquateEmpty()); diff != "" {
+					t.Errorf("Extract() inventory mismatch (-want +got):\n%s", diff)
+				}
+
+				// If testing already-processed scenario, call Extract again
```
Comment (erikvarga) [inline extractor/filesystem/bazel/aspect/aspect_test.go]: Let's make a separate test function for this instead of adding it here since this is only ever called once.

**Comment by neilu-jain** [comment]
updated

**Review by erikvarga — APPROVED** [review]

**Comment by erikvarga** [comment]
@neilu-jain it looks like some additional internal checkers found some formatting/docstring issues with the .bzl file. Can you update the file to address them? I 'cc-d you on the import CL with the failure message.
